capstone/src/main/java/com/google/sps/servlets/TagsTFIDFServlet.java

```diff
@@ -144,21 +192,59 @@ public void calculateTFIDF(LinkedHashMap<Long, LinkedHashMap<String, Integer>> g
     }
   }
 
+  /**
+   * Weights the tf-idf score of a given ngram by taking into account its length.
+   */
+  private double weightScore(double score, String ngram) {
+    double wordsWeight = (ngram.split(" ").length / 3.1);
+    double lengthWeight = ((double) ngram.length() / longestNgramLength);
+
+    double totalWeight = 1 + wordsWeight + lengthWeight;
+    score *= totalWeight;
+
+    return score;
+  }
+
   /**
    * Given a priority queue of tags, stores the top 3 tag values in the Group's datastore.
    */
   public void putTagsInDatastore(PriorityQueue<Tag> tagQueue, long groupId, 
       HttpServletResponse response) throws IOException {
-
+    
     ArrayList<Tag> topTags = new ArrayList<>();
     int tagsLength = (tagQueue.size() >= 3) ? 3 : tagQueue.size();
     for (int i = 0; i < tagsLength; i++) {
-      topTags.add(tagQueue.poll());
+      Tag next = tagQueue.poll();
+      if (next != null) {
+        if (checkIfDuplicate(topTags, next)) {
+          i--;
```

Comment:
shouldn't this be a no-op instead? if you do i--, don't you end up checking the same value repeatedly by doing i--, and then in the for loop i++?

```java
if (!checkIfDuplicate(topTags, next)) {
  topTags.add(next);
}
```

Reply:
I guess high level, I'm not sure if this is the direction you want

You're not allowing any overlap between any terms in ngrams, so `harry and hermione` and `potter and the` are invalid ngrams because they both use the word "and"

Reply:
No, earlier I call tagQueue.poll() which always retrieves and removes the first element in the queue.
The idea is that I want to loop through 3 times, and end up with three items added to the list. If I didn't decrement i, I might only end up with one or two items in the list.

Reply:
That's a good point about small words overlapping causing issues. I'm not really sure how else to best check for duplicate terms though. 
Currently for the case you described, if `harry and hermione` are added first because it has the highest score, and `potter and the` has the next highest score, `potter and the` would be counted as a duplicate and wouldn't be added, but `harry and hermione` would remain.

Maybe to mitigate against cases like that, I could say that if both terms I'm checking have two or more words, only mark it as a duplicate if two or more of those words are a match with the words in a term currently in the list. Here's what the results would be:

Currently in list         | Word to be checked  | Duplicate?
`harry and hermione` | `potter and the` | no
`harry and hermione` | `ron and hermione` | yes
`harry and hermione` | `harry` | yes
`harry and hermione` | `harry and` | yes
`harry and` | `harry and hermione` | yes
`and hermione` | `ron and hermione` | yes
`and hermione` | `ron and harry` | no
`and hermione` | `and harry` | no

 Not sure if it's an ideal solution. I'm brainstorming other possibilities. Thoughts?

Reply:
The answer MAY be not dedup-ing but adding another weight that measures duplicate results. You have your sliders of [0-1) that you can add a third one for "how duplicate" a string is?

I actually think I like your outcomes for the solution you listed though, I think all things I want to be no were no, and all things I want to be yes were yes.

You'd imagine in the `harry and hermione` | `ron and hermione` the winning ngram would be `and hermione`, due to it being significantly more prevelant than `harry and hermione` || `ron and hermione`

I think the relationship we're looking for is proportional, where `harry and hermione` should beat `and hermione` if it occurs > 2/3 of the time where 2 is the length of the 2gram and 3 is the length of the 3gram. Thoughts?

Reply:
This would also mean that `Disney Plus` would only beat `Disney` if more than half of the time `Disney` was paired with `Plus`

Reply:
This step would happen after scoring and ranking so realistically `harry and hermione` and `ron and hermione` _probably_ wouldn't both show up. If they did, `and hermione` would definitely already be in the list, negating both of them. 

I think adding another "slider" wouldn't be very efficient since then I would have to do a similar checkHowDuplicate calculation for every single ngram. For now, I'm only doing doing dedup logic for the top handful of scores, but by introducing it as a weight I would have to calculate how duplicate a term is for hundreds or thousands of ngrams.

I think the proportional solution makes sense in theory but I am not sure how to keep track of that. By the time I'm doing dedup I've already ranked everything by its frequency, length, etc. So just comparing the scores to get relative frequency wouldn't work because they're additionally weighted by n and the string length.